File: 贪心/53最大子序和.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res = Sum = nums[0]
        # Sum为i-1结尾的最大连续子数组。i时Sum+=nums[i]，
        # 但如果Sum<0就当作0，最大连续子数组就是nums[i]本身
        for i in range(1, len(nums)):
            if Sum < 0: Sum = 0
            Sum += nums[i]
            res = max(res, Sum)
        return res

File: 动态规划/53最大子序和.java
class Solution {
    public int maxSubArray(int[] nums) {
        int[] dp = new int[nums.length]; // 以nums[i]结尾的最大子序和
        dp[0] = nums[0];
        int max = dp[0];
        for(int i = 1; i < nums.length; i++) {
            dp[i] = Math.max(0, dp[i-1]) + nums[i];
            max = Math.max(max, dp[i]);
        }
        return max;
    }
}
